src/Components/Blazor/WebAssembly.Authentication/src/AuthenticationManager.cs

```diff
@@ -0,0 +1,446 @@
+                switch (result.Status)
+                {
+                    case RemoteAuthenticationStatus.Redirect:
+                        break;
+                    case RemoteAuthenticationStatus.Success:
+                        await NavigateToReturnUrl(returnUrl);
+                        break;
+                    case RemoteAuthenticationStatus.OperationCompleted:
+                        break;
+                    case RemoteAuthenticationStatus.Failure:
+                        var uri = Navigation.ToAbsoluteUri($"{ApplicationPaths.LogoutFailedPath}?message={result.ErrorMessage}").ToString();
+                        Navigation.NavigateTo(uri);
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Invalid authentication result status '{result.Status ?? "(null)"}'.");
+                }
+            }
+        }
+
+        private async Task ProcessLogoutCallback()
+        {
+            var result = await AuthenticationService.CompleteSignOutAsync(new RemoteAuthenticationContext<TAuthenticationState> { Url = Navigation.Uri });
+            switch (result.Status)
+            {
+                case RemoteAuthenticationStatus.Redirect:
+                    // There should not be any redirects as the only time completeAuthentication finishes
+                    // is when we are doing a redirect sign in flow.
+                    throw new InvalidOperationException("Should not redirect.");
+                case RemoteAuthenticationStatus.Success:
+                    await NavigateToReturnUrl(GetReturnUrl(result.State, Navigation.ToAbsoluteUri(ApplicationPaths.LogoutSucceededPath).ToString()));
+                    break;
+                case RemoteAuthenticationStatus.OperationCompleted:
+                    break;
+                case RemoteAuthenticationStatus.Failure:
+                    var uri = Navigation.ToAbsoluteUri($"{ApplicationPaths.LogoutFailedPath}?message={result.ErrorMessage}").ToString();
+                    Navigation.NavigateTo(uri);
+                    break;
+                default:
+                    throw new InvalidOperationException($"Invalid authentication result status '{result.Status ?? "(null)"}'.");
+            }
+        }
+
+        private string GetReturnUrl(TAuthenticationState state, string defaultReturnUrl = null)
+        {
+            var fromQuery = GetParameter("returnUrl");
+            if (!string.IsNullOrWhiteSpace(fromQuery) && !fromQuery.StartsWith(Navigation.BaseUri))
+            {
+                // This is an extra check to prevent open redirects.
+                throw new InvalidOperationException("Invalid return url. The return url needs to have the same origin as the current page.");
+            }
+
+            return state?.ReturnUrl ?? fromQuery ?? defaultReturnUrl ?? Navigation.BaseUri;
+        }
+
+        private string GetParameter(string key)
```

Comment:
Would it be OK to factor this out into a separate class? It looks like something that would be generally useful, and would make sense to have its own separate set of tests.

Reply:
Also out of interest, does the logic in this method come from somewhere else, or was it written freshly for this scenario?

Reply:
I'm fine with trusting the implementation and not adding extra tests, but would still have a preference for pulling it out into a separate `internal` class and putting on a comment block that explains where it comes from.

That will make it much easier to reason about either (a) reusing it elsewhere in the future or (b) swapping in a different implementation in the future, if we have need for either.

Reply:
Moved this into a separate internal class.